solvers/ADMM.py

```diff
@@ -33,22 +34,21 @@ def set_objective(self, A, reg, y, c, delta, data_fit):
 
     def run(self, callback):
         p = self.A.shape[1]
-        data = np.array([-np.ones(p), np.ones(p)])
-        diags = np.array([0, 1])
-        D = spdiags(data, diags, p-1, p)
         u = self.c * np.ones(p)
         z = np.zeros(p - 1)
         mu = np.zeros(p - 1)
         gamma = self.gamma
-        AtA = self.A.T @ self.A
-        DtD = D.T @ D
         Aty = self.A.T @ self.y
-        A_tmp = np.linalg.pinv(AtA + gamma * DtD)
+        AtA_gDtD = LinearOperator(shape=(p, p),
+                                  matvec=lambda x: self.A.T @ (self.A @ x) -
+                                  gamma * np.diff(np.diff(x),
+                                                  append=0, prepend=0))
 
         while callback(u):
             z_old = z
-            u = np.ravel(A_tmp @ (Aty + np.diff(mu, append=0, prepend=0)
-                                  - gamma * np.diff(z, append=0, prepend=0)))
+            u_tmp = Aty + np.diff(mu, append=0, prepend=0) - \
+                gamma * np.diff(z, append=0, prepend=0)
```

Comment:
Aviod using `\` in the future, use parenthesis if you need multi-line expressions.